cfme/login.py

```diff
@@ -33,6 +28,21 @@ def __init__(self, username=None, password=None, full_name=None):
         self.full_name = full_name
         self.password = password
         self.username = username
+        self._restore_user = None
+
+    def __enter__(self):
+        if not store.user or store.user.username != self.username:
+            logger.info('Switching to new user: {}'.format(self.username))
+            self._restore_user = store.user
```

Comment:
Now that this exists outside of the abstract, the notion of making this reentrant is much easier to explain.

What would happen if, inside this context, someone entered another user context? As far as I can tell, it would work. But then, what if someone tries to enter *that same user's* context? As it stands, I think our `_restore_user` will be lost. We should handle this behavior by either accounting for it or forbidding it, and given those options I think it's more pythonic to allow than to deny.

```
if self._restore_user != store.user:
  self._restore_user = store.user
  logout()
  # set_user should be abolished in favor of a property setter
  store.user = self
# else do nothing, we're already in this user's context and this is a no-op
```

And on the exit side:
```
if self._restore_user != store.user:
  logout()
  # set_user should be abolished in favor of a property setter
  store.user = self._restore_user
  self._restore_user = None
# else do nothing, we're exiting this context into the context of *the same user* and this is a no-op
```

Now we should be able to safely nest different user contexts, as well as have a given user's context be reentrant, provided we properly implement `__eq__` to allow for reasonable user equality checking.